parser.go

```diff
@@ -244,7 +195,18 @@ func (p *Parser) parseStatements(doParse func()) {
 //
 // ================================================================================
 
-func (p *Parser) parseQueryStatement() *ast.QueryStatement {
+func (p *Parser) parseQueryStatement() (stmt *ast.QueryStatement) {
+	l := p.Lexer.Clone()
+	defer func() {
+		if r := recover(); r != nil {
+			// When parsing is failed on tryParseHint or tryParseWith, the result of these methods are discarded
```

Comment:
I have concerning discarding hints and CTEs means input statement can't be recovered by `SQL()`. I think it don't satisfy users needs(parse-unparse should be safe)

Reply:
Parsing results are discarded, but the contents (tokens) are preserved. Therefore, your concern is not a problem.

Reply:
I mean, parse-unparse is still safe after this PR.

Reply:
Thank you for explanation.
I understood all `handleParse*Error` will eat from `l.Token.Pos` to their stop position.